File: Everything-app-backend/everything-app/src/main/java/com/Finn/everything_app/model/FlashcardReview.java
package com.Finn.everything_app.model;

import jakarta.persistence.*;
import lombok.Data;
import java.time.LocalDateTime;

/**
 * Ein einzelnes Review — das Protokoll hinter den Lernstatistiken.
 *
 * Ohne diese Zeilen gäbe es kein „heute X Karten gelernt", keine Retention-Kurve und kein
 * Rückgängig: die Karte selbst kennt nur ihren aktuellen Zustand, nicht ihren Weg dorthin.
 * Kostet einen Insert pro Bewertung.
 *
 * Das Intervall wird VOR und NACH der Bewertung festgehalten, weil erst die Differenz zeigt,
 * ob eine Karte gewachsen oder zurückgefallen ist.
 */
@Entity
@Table(name = "flashcard_reviews")
@Data
public class FlashcardReview {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "flashcard_id", nullable = false)
    private Flashcard flashcard;

    // Redundant zu flashcard.deck.user, aber die Auswertung filtert immer nach Nutzer und
    // Zeitraum — ohne die Spalte hinge an jeder Statistik ein Doppel-Join.
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 10)
    private ReviewRating rating;

    @Column(name = "reviewed_at", nullable = false)
    private LocalDateTime reviewedAt;

    @Column(name = "interval_days_before")
    private Double intervalDaysBefore;

    @Column(name = "interval_days_after")
    private Double intervalDaysAfter;

    @Column(name = "ease_after")
    private Double easeAfter;

    @PrePersist
    protected void onCreate() {
        if (reviewedAt == null) {
            reviewedAt = LocalDateTime.now();
        }
    }
}

File: Everything-app-frontend/everything_app/lib/models/flashcard_review.dart
import '../utils/anki_scheduler.dart';

/// Eine protokollierte Bewertung.
///
/// Die Karte selbst kennt nur ihren aktuellen Zustand. Erst dieses Protokoll beantwortet
/// „wie viel habe ich heute gelernt" und zeigt über [intervalDaysBefore]/[intervalDaysAfter],
/// ob eine Karte gewachsen oder zurückgefallen ist.
class FlashcardReview {
  final String id;
  final String flashcardId;
  final String deckId;
  final ReviewRating rating;
  final DateTime reviewedAt;
  final double intervalDaysBefore;
  final double intervalDaysAfter;

  const FlashcardReview({
    required this.id,
    required this.flashcardId,
    required this.deckId,
    required this.rating,
    required this.reviewedAt,
    this.intervalDaysBefore = 0,
    this.intervalDaysAfter = 0,
  });

  factory FlashcardReview.fromJson(Map<String, dynamic> json) => FlashcardReview(
        id: json['id'].toString(),
        flashcardId: json['flashcardId']?.toString() ?? '',
        deckId: json['deckId']?.toString() ?? '',
        rating: _ratingFrom(json['rating']),
        reviewedAt: json['reviewedAt'] != null
            ? DateTime.parse(json['reviewedAt'])
            : DateTime.now(),
        intervalDaysBefore: (json['intervalDaysBefore'] as num?)?.toDouble() ?? 0,
        intervalDaysAfter: (json['intervalDaysAfter'] as num?)?.toDouble() ?? 0,
      );

  /// Das Backend schreibt AGAIN/HARD/GOOD/EASY, das Dart-Enum heißt again/hard/good/easy.
  static ReviewRating _ratingFrom(dynamic raw) {
    final name = raw?.toString().toLowerCase();
    return ReviewRating.values.firstWhere(
      (r) => r.name == name,
      orElse: () => ReviewRating.good,
    );
  }

  /// Die Karte ist bei dieser Bewertung zurückgefallen.
  bool get isLapse => rating == ReviewRating.again;
}
